build-tools/packages/build-tools/src/repoPolicyCheck/handlers/pnpm.ts

```diff
@@ -0,0 +1,70 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import fs from "fs";
+import path from "path";
+
+import { PackageJson } from "../../common/npmPackage";
+import { Handler, readFile } from "../common";
+
+const match = /(?:^|\/)pnpm-lock\.yaml$/i;
+export const handlers: Handler[] = [
+	{
+		name: "pnpm-npm-package-json-preinstall",
+		match,
+		handler: (file) => {
+			const dirname = path.dirname(file);
+			const packageJsonFile = path.join(dirname, "package.json");
+			let json: PackageJson;
+			try {
+				json = JSON.parse(readFile(packageJsonFile));
+			} catch (err) {
+				return "Error parsing JSON file: " + packageJsonFile;
+			}
+
+			const script: string | undefined = json.scripts?.preinstall;
+			if (script) {
+				const matchResult = script.match(/^node ((?:\.\.\/)*)scripts\/only-pnpm.cjs/);
+				if (matchResult) {
+					const onlyPnpmPath = path.join(
+						dirname,
+						`${matchResult[1] ?? ""}scripts/only-pnpm.cjs`,
+					);
+					if (!fs.existsSync(onlyPnpmPath)) {
+						return `pnpm enforcement preinstall script "${onlyPnpmPath}" does not exist`;
+					}
+				} else {
+					return `Invalid pnpm enforcement preinstall script "${script}" in package.json`;
+				}
+			} else {
+				return `package.json missing pnpm enforcement preinstall script`;
+			}
+			return undefined;
+		},
+	},
+	{
+		name: "pnpm-lock-workspace",
```

Comment:
```suggestion
		// A package or workspace that uses pnpm must have a pnpm-workspace.yaml file, even if it is just a single package.
		// This is needed because we have a workspace in the root so independent packages need one at the package level to override it.
		name: "pnpm-lock-workspace",
```